### ISSUE TITLE
Methods that return shared references are at risk for undefined behaviour

### ISSUE THREAD

**Issue by mmomtchev** [mmomtchev/nobind#1]
Consider this chainable C++ method:
```cpp
class Chainable {
public:
  Chainable();
  Chainable &chainable() { return *this; }
};
```

In Javascript it could be used as:
```js
const obj = new Chainable;
obj.chainable().chainable().chainable();
```

which is ok, however doing this:
```js
let obj = new Chainable;
obj = obj.chainable().chainable().chainable();
```

Will replace the original `obj` which owned its C++ object with the shared copy obtained from the last invocation of the chainable method. When the original copy gets expired by the GC, this won't invalidate the shared reference which will be a dangling pointer.

Alas, there is no simple solution (the right solution would be an object store), and given that the current goal is to limit the code complexity, it would stay like this for 1.0.

Short version: be extra careful with functions that return shared references to dynamic objects
